barretenberg/cpp/src/barretenberg/honk/pcs/zeromorph/zeromorph.hpp

```diff
@@ -0,0 +1,357 @@
+#pragma once
+#include "barretenberg/polynomials/polynomial.hpp"
+
+/**
+ * @brief
+ *
+ */
+namespace proof_system::honk::pcs::zeromorph {
+
+/**
+ * @brief Prover for ZeroMorph multilinear PCS
+ *
+ * @tparam Curve
+ */
+template <typename Curve> class ZeroMorphProver_ {
+    using Fr = typename Curve::ScalarField;
+    using Commitment = typename Curve::AffineElement;
+    using Polynomial = barretenberg::Polynomial<Fr>;
+
+    // TODO(#742): Set this N_max to be the number of G1 elements in the mocked zeromorph SRS once it's in place. (Then,
+    // eventually, set it based on the real SRS).
+    static const size_t N_max = 1 << 10;
+
+  public:
+    /**
+     * @brief Compute multivariate quotients q_k(X_0, ..., X_{k-1}) for f(X_0, ..., X_{d-1})
+     * @details Given multilinear polynomial f = f(X_0, ..., X_{d-1}) for which f(u) = v, compute q_k such that:
+     *
+     *  f(X_0, ..., X_{d-1}) - v = \sum_{k=0}^{d-1} (X_k - u_k)q_k(X_0, ..., X_{k-1})
+     *
+     * The polynomials q_k can be computed explicitly as the difference of the partial evaluation of f in the last k
+     * variables at, respectively, u'' = (u_k + 1, u_{k+1}, ..., u_{n-1}) and u' = (u_{n-k-1}, ..., u_{n-1}). I.e.
+     *
+     *  q_k(X_0, ..., X_{k-1}) = f(X_0,...,X_{k-1}, u'') - f(X_0,...,X_{k-1}, u')
+     *
+     * TODO(#739): This method has been designed for clarity at the expense of efficiency. Implement the more efficient
+     * algorithm detailed in the latest versions of the ZeroMorph paper.
+     * @param input_polynomial Multilinear polynomial f(X_0, ..., X_{d-1})
+     * @param u_challenge Multivariate challenge u = (u_0, ..., u_{d-1})
+     * @return std::vector<Polynomial> The quotients q_k
+     */
+    static std::vector<Polynomial> compute_multilinear_quotients(std::span<Fr> input_poly, std::span<Fr> u_challenge)
+    {
+        auto input_polynomial = Polynomial(input_poly);
+        size_t N = input_polynomial.size();
+        size_t log_N = numeric::get_msb(N);
+
+        // Define the vector of quotients q_k, k = 0, ..., log_n-1
+        std::vector<Polynomial> quotients;
+        for (size_t k = 0; k < log_N; ++k) {
+            size_t size = 1 << k;
+            quotients.emplace_back(Polynomial(size)); // degree 2^k - 1
+        }
+
+        // Compute the q_k in reverse order, i.e. q_{n-1}, ..., q_0
+        for (size_t k = 0; k < log_N; ++k) {
+            // Define partial evaluation point u' = (u_{n-k-1}, ..., u_{n-1})
+            auto partial_size = static_cast<std::ptrdiff_t>(k + 1);
```

Comment:
fixed, thanks

Reply:
ah ok,  now i understand why the partial size is k+1  because you allocate the size of the vector as n-k-1. hm, maybe partial_size should be called something else because i interpreted it as `u_partial`

Reply:
maybe evaluation_point_size

Reply:
yeah it's slightly confusing, I wasn't totally happy with this code. To be clear though, `partial_size` is the size of `u_partial`